python/tests/hypothesis/arcticdb/test_sort_merge.py

```diff
@@ -8,64 +7,316 @@
+    if not IS_PANDAS_TWO:
+        # Due to https://github.com/man-group/ArcticDB/blob/7479c0b0caa8121bc2ca71a73e29769bbc41c66a/python/arcticdb/version_store/_normalization.py#L184
+        # we change the dtype of empty float columns. This makes hypothesis tests extremely hard to write as we must
+        # keep addional state about is there a mix of empty/non-empty float columns in the staging area, did we write
+        # empty float column (if so it's type would be object). These edge cases are covered in the unit tests.
+        index = hs_pd.indexes(dtype="datetime64[ns]", min_size=1)
+    else:
+        index = hs_pd.indexes(dtype="datetime64[ns]")
+    return draw(hs_pd.data_frames(columns, index=index))
+
+def assert_equal(left, right, dynamic=False):
+    """
+    The sorting Arctic does is not stable. Thus when there are repeated index values the
+    result from our sorting and the result from Pandas' sort might differ where the index
+    values are repeated.
+    """
+    if any(left.index.duplicated()):
+        assert left.index.equals(right.index), f"Indexes are different {left.index} != {right.index}"
+        assert set(left.columns) == set(right.columns), f"Column sets are different {set(left.columns)} != {set(right.columns)}"
+        assert left.shape == right.shape, f"Shapes are different {left.shape} != {right.shape}"
+        left_groups = left.groupby(left.index, sort=False).apply(lambda x: x.sort_values(list(left.columns)))
+        right_groups = right.groupby(right.index, sort=False).apply(lambda x: x.sort_values(list(left.columns)))
+        assert_frame_equal(left_groups, right_groups, check_like=True, check_dtype=False)
+    else:
+        assert_frame_equal(left, right, check_like=True, check_dtype=False)
+
+def assert_cannot_finalize_without_staged_data(lib, symbol, mode):
+    with pytest.raises(UserInputException) as exception_info:
+        lib.sort_and_finalize_staged_data(symbol, mode=mode)
+    assert "E_NO_STAGED_SEGMENTS" in str(exception_info.value)
+  
+def assert_nat_is_not_supported(lib, symbol, mode):
+    with pytest.raises(UnsortedDataException) as exception_info:
+        lib.sort_and_finalize_staged_data(symbol, mode=mode)
+    assert "E_UNSORTED_DATA" in str(exception_info.value)
+
+def assert_staged_columns_are_compatible(lib, symbol, mode):
+    with pytest.raises(SchemaException):
+        lib.sort_and_finalize_staged_data(symbol, mode)
+
+def has_nat_in_index(segment_list):
+    return any(pd.NaT in segment.index for segment in segment_list)
+
+def merge_and_sort_segment_list(segment_list, int_columns_in_df=None):
+    merged = pd.concat(segment_list)
+    # pd.concat promotes dtypes. If there are missing values in an int typed column
+    # it will become float column and the missing values will be NaN.
+    int_columns_in_df = int_columns_in_df if int_columns_in_df else []
```

Comment:
On a second thought this is not quite reliable especially with dynamic schema. Columns can start as int then get promoted to float. I think if we can make sparse maps work this check will work. Until then I'm thinking of replacing all nans with 0 in both source and expected.